cloudinit/sources/DataSourceAzure.py

```diff
@@ -660,10 +691,249 @@ def setup(self, is_new_instance):
+        # we do not yet know which is the primary NIC, we
+        # do dhcp on each NIC until we find it.
+        with events.ReportEventStack(
+                name="obtain-dhcp-lease",
+                description="obtain dhcp lease",
+                parent=azure_ds_reporter):
+            # Do dhcp with this nic
+            dhcp_ctx = EphemeralDHCPv4(
+                iface=ifname,
+                dhcp_log_func=dhcp_log_cb)
+            dhcp_ctx.obtain_lease()
+
+        # Primary nic detection will be optimized in the
+        # future. The fact that primary nic is being
+        # attached first helps here. Otherwise each nic
+        # could add several seconds of delay.
+        try:
+            imds_md = get_metadata_from_imds(
+                ifname,
+                5,
+                metadata_type.network)
+            if imds_md is not None:
+                LOG.info("%s is the primary nic", ifname)
+                is_primary = True
+
+                # If primary, set ephemeral dhcp ctx so we
+                # can report ready
+                self._ephemeral_dhcp_ctx = dhcp_ctx
+
+                # Set the expected nic count based on the
+                # response received
+                expected_nic_count = len(
+                    imds_md['interface'])
+                LOG.info("Expected nic count: %d",
+                         expected_nic_count)
+        except Exception as e:
+            LOG.warning(
+                "Failed to get network metadata using nic "
+                "%s. Attempt to contact IMDS failed with "
+                "error %s. Assuming this is not the "
+                "primary nic.", ifname, e)
+        finally:
+            # If we are not the primary nic, then clean the
+            # dhcp context.
+            if imds_md is None:
+                dhcp_ctx.clean_network()
+
+        return is_primary, expected_nic_count
+
+    @azure_ds_telemetry_reporter
+    def _wait_for_hot_attached_nics(self, nl_sock):
+        """Wait until all the expected nics for the vm are hot-attached.
+        The expected nic count is obtained by requesting the network metadata
+        from IMDS.
+        """
+        LOG.info("Waiting for nics to be hot-attached")
+        try:
+            # Wait for nics to be attached one at a time, until we
+            # know for sure that all nics have been attached.
+            nics_found = []
+            primary_nic_found = False
+            expected_nic_count = -1
+
+            # Wait for netlink nic attach events.
+            # After the first nic is attached, we are already in the
+            # customer vm deployment path and so everything from then
+            # on should happen fast and avoid unnecessary delays
+            # wherever possible.
+            while True:
+                ifname = netlink.wait_for_nic_attach_event(nl_sock,
+                                                           nics_found)
```

Comment:
This is my last comment on diagnostics. Realized this is needed otherwise there is no telemetry if there are delays in nic at
```suggestion
                with events.ReportEventStack(
                        name="wait-for-nic-attach",
                        description="wait for nic attach after %d nics have already been attached" % len(nics_found),
                        parent=azure_ds_reporter):
                    ifname = netlink.wait_for_nic_attach_event(nl_sock, nics_found)
```